src/app/SIPUserAgents/Behaviours/SIPAuthChallenge.cs

```diff
@@ -0,0 +1,36 @@
+﻿//-----------------------------------------------------------------------------
+// Filename: SIPAuthChallenge.cs
+//
+// Description: Common logic when having to add auth to SIP Requests
+//
+// History:
+// 10 May 2021 : Created 
+//
+// License: 
+// BSD 3-Clause "New" or "Revised" License, see included LICENSE.md file.
+//-----------------------------------------------------------------------------
+using System.Linq;
+using SIPSorcery.SIP;
+
+
+namespace SIPSorcery.App.SIPUserAgents.Behaviours
+{
+    public static class SIPAuthChallenge
+    {
+        public static SIPRequest AddAuthenticationHeaderToRequest(SIPRequest previousSipRequest, SIPResponse sipResponse, string username, string password)
+        {
+            // Resend Invite with credentials.
+            SIPAuthorisationDigest authRequest = sipResponse.Header.AuthenticationHeader.SIPDigest;
+            SIPURI contactUri = sipResponse.Header.Contact.Any() ? sipResponse.Header.Contact[0].ContactURI : sipResponse.Header.From.FromURI;
+            
+            authRequest.SetCredentials(username, password, contactUri.ToString(), previousSipRequest.Method.ToString());
+
```

Comment:
The URI used in the SetCredentials call should be the request URI, i.e. previousSipRequest.URI, not the ContactURI.

Reply:
Indeed, considering we are attempting to re-send the request and not the response

Reply:
Considering this, then it means that the we can't use the alternative Header.From.FromURI either, even if it is from the previousSipRequest. This could be wrong if the From is not the actual originator of the request. I think we might need to let the caller select depending on context what is the contact URI in case the original sip request does not have any

Reply:
Nvmm just went through RFC 3265: 

> The From header field indicates the logical identity of the initiator of the request

So it seems ok to use the From header. I mistakenly thought it was based on the initiator of the original call

Reply:
The URI used in the digest needs to be the request URI, that is the URI on the first line of the SIP request, see https://datatracker.ietf.org/doc/html/rfc3261#section-22.4. It can't be the Contact of From header which could both be different to the request URI.

Reply:
I see it now.  Based on what [https://datatracker.ietf.org/doc/html/rfc2617#section-3.2.2](https://datatracker.ietf.org/doc/html/rfc2617#section-3.2.2) says, since SIP Auth follows that definition